unmerge fix-frontend pr and continues working

idkjs opened the pull request:
Reverts project to ec70eb365ef1dfea1f54f8a64a9143e45bfbee47 with hard reset and adds changes from fix-frontend erroneously merged pr. 

The effect here is to unmerge that pr and finish what we want to do here. Does this accomplish what we talked about.

davesnx commented on Makefile:
```diff
@@ -12,6 +12,14 @@ help:
 
+.PHONY: db
+db: ## Start the database with dune exec
+	dune exec counter
```
That's not the command to run the db...

idkjs replied:
Right, as I noted in the previous PR, this is a hack for it to work while developing. I guess we need to fix it. Running that command gets you:
```sh
~/Github/CAAS/backend-only fix-frontend2 !2                                                                                             13s  3.0.1
❯ make build
opam exec -- dune build @all

~/Github/CAAS/backend-only fix-frontend2 !2                                                                                              6s  3.0.1
❯ make db
dune exec counter
counter: [INFO] Starting Opium on 0.0.0.0:3003
```

Its the equivalent of `make run`...which apparently also works.
```sh
❯ make run
opam exec -- dune exec counter
counter: [INFO] Starting Opium on 0.0.0.0:3003
```

I will remove it. Thanks.

davesnx commented on Makefile:
```diff
@@ -12,6 +12,14 @@ help:
+
+.PHONY: front
+front: ## Start the front with yarn start
+	yarn start
```
Frontend commands shoudn't be on the makefile

idkjs replied:
Removed.

davesnx commented on Makefile:
```diff
@@ -26,7 +34,7 @@ install: ## Install the dependencies
 .PHONY: init
 init: ## Create a local opam switch and install deps
-	opam switch create . 4.10.0 --deps-only --locked
+	# opam switch create . 4.12.0 --deps-only --locked
```
init should create the switch, install ocaml-lsp and maybe not do the make install, but please don't change it in this PR

idkjs replied:
reverted.

davesnx commented on backend/db/fixtures.sql:
```diff
@@ -1,3 +1,11 @@
+CREATE EXTENSION IF NOT EXISTS "uuid-ossp";
+DROP TABLE IF EXISTS "counters";
+CREATE TABLE counters (
+  id TEXT PRIMARY KEY DEFAULT extensions.uuid_generate_v4(),
+  name TEXT NOT NULL,
+  value INTEGER
+);
+
```
I have seen similar changes over and over, could you try to explain why those are necessary? 

I assume we got everything working fine last changes

idkjs replied:
I dont know why `uuid_generate_v4()` doesnt work on its own anymore. I remember it worked previously. After researching it, this is how I got it to work. The only thing I can think of was that I got an error saying that `uuid-ossp` was not installed on my machine. I do not recall how it was installed previously. I then installed with `brew install uuid-ossp` and it worked.

What do you think?

davesnx replied:
Fixtures shoudn't need migrations, not sure what you mean with uuid_generate_v4 here

idkjs replied:
Also, the db scripts in the make file dont work for me so to get `uuid-ossp` set up on postgres we have to run it with pure sql on the local machine. I should put in another script for local development. I just did that.

So how do we seed the postgres db with our `fixtures.sql`? Seems like the db scripts are working now.

davesnx replied:
Can we remove this migration on the fixtures then?

idkjs replied:
Im not sure what we are talking about here. The word `migrations` appears in `fixtures.sql` because I copied from your code. My understanding is that this is a seed script so we can test some http calls. Maybe that is a migration. The point of the `fixtures.sql` script is to have some fixed data that we can test against. Everything in the file seems necessary for that. 

As I say that, do you mean you want to move:
```sql
-- migrations to seed db
INSERT INTO counters (name, value)
VALUES ('How many tacos did I eat 🌮 ?', 0),
          ('Counting', 0),
          ('Number of coffes', 2),
          ('Compete', 100);
``` 
into `backend/db/migrations.re`?

davesnx commented on bsconfig.json:
```diff
@@ -25,14 +25,15 @@
   "warnings": {
     "number": "-44-45",
-    "error": "+8+26+27+101+103"
+    "error": "+26+27+101+103"
```
Why?

idkjs replied:
If we have the `+8` set as an error dont compile with the following error:
```sh
  /Users/mando/Github/CAAS/update-frontend/frontend/components/UI.re:104:5-120:5

  102 │ module Spacer = [%styled.div
  103 │   (~top=0, ~left=0, ~bottom=0, ~right=0, ~all=0) => [|
  104 │     switch (
  105 │       unitize(top),
    . │ ...
  119 │     | (_, _, _, _, v) when v != 0 => CssJs.unsafe("margin", v |> stri
        ngToPx)
  120 │     },
  121 │   |]
  122 │ ];

  this pattern-matching is not exhaustive.
All clauses in this pattern-matching are guarded.
```

I have changed it to a warning so we can still see what is happening and also get compilation. I tried to fix it but was unable to. 

Do you know how to fix this issue?

davesnx replied:
I will fix this, can u disabled the warning just for this component?

idkjs replied:
Done.

davesnx commented on frontend/atd/Counter_bs.ml:
```diff
@@ -0,0 +1,153 @@
+(* Auto-generated from "Counter.atd" *)
```
auto-generated code from atd shoudn't be part of the repository

idkjs replied:
Added `.gitignore` to directory.

davesnx commented on frontend/components/Api.re:
```diff
@@ -0,0 +1,107 @@
+let baseUrl = "http://127.0.0.1:3003/";
```
Could you please move back the Api to frontend/Api.re? Have nothing to do with components

davesnx commented on frontend/components/App.re:
```diff
@@ -1,30 +1,38 @@
 
-module H1 = [%styled.h1 {|
+module H1 = [%styled.h1
+  {|
```
Could you disabled refmt for this project and keep the indentation as before?

idkjs replied:
Done.

davesnx commented on frontend/index.re:
```diff
@@ -1,6 +1,7 @@
-// include GlobalStyles;
+include GlobalStyles;
 
 switch (ReactDOM.querySelector("#root")) {
 | Some(root) => ReactDOM.render(<App />, root)
 | None => Js.log("Coudn't find #root element to mount the React app.")
 };
+// ReactDOMRe.renderToElementWithId(<App />, "root");
```
Why?

idkjs replied:
The now removed commented out code throws this error.
```sh
 7 │ ReactDOMRe.renderToElementWithId(<App />, "root");

  The value renderToElementWithId can't be found in ReactDOMRe
```

davesnx commented on package.json:
```diff
@@ -7,28 +7,32 @@
+    "bs:build": "NINJA_ANSI_FORCED=1 bsb -make-world",
+    "bs:watch": "NINJA_ANSI_FORCED=1 bsb -make-world -w",
+    "bs:clean": "bsb -clean-world",
+    "start": "yarn clean; concurrently --names re,db,dev,js --kill-others yarn:bs:watch yarn:dev-server yarn:bundle:watch",
```
I'm fine using yarn if there's a good reason, but now in this PR we have both lock files. Please choose one and update all references to it.

idkjs replied:
Done. Reverted to watchexec and used npm throughout. I switched to npm because it was working while yarn was throwing some error I do not recall at this point in time. If you prefer yarn I can go see what that was or if it still exists.

idkjs replied:
I reverted back to yarn in d2ef60c4a8838ecdbc3c0b23ed57a0fe31927a2d because I checked it and it works and its what you had originally in addition to the scripts where not working as expected.

davesnx commented:
I left a bunch of comments in this PR, but the main issue that I see here is that there's no "Why" or "What" is this PR. 

Happy to review half-backed PRs (probably better turn it into drafts when are a work in progress), but for PRs that can be reviewed would be highly beneficial to have a clear description of changes.

Let me know if that make sense.

idkjs commented on package.json:
```diff
@@ -7,28 +7,32 @@
-    "bundle:prod": "npm run bundle --minify --define:process.env.NODE_ENV='\"production\"'",
-    "clean": "rm -rf www/js; npm run bs:clean",
+    "bs:build": "NINJA_ANSI_FORCED=1 bsb -make-world",
+    "bs:watch": "NINJA_ANSI_FORCED=1 bsb -make-world -w",
```
Done.

idkjs commented:
> Let me know if that make sense.

Absolutely makes sense and I should have made it a draft. Thanks for the feedback.

idkjs commented:
I have addressed all the comments. To run this now you can run `make init` then `make run` for the db and `yarn && yarn start` for the frontend. I need to figure out how to seed the docker db with the fixtures.

davesnx commented on frontend/Api.re:
```diff
@@ -3,45 +3,105 @@ let countersUrl = baseUrl ++ "counters";
 
-type t = RemoteData.t(Counter_t.counters);
-
+// type t = RemoteData.t(Counters.Counter_t.counters);
+type t = RemoteData.t(array(Counter_t.t));
```
?

idkjs replied:
Removed comment. Thank you.

davesnx commented on frontend/atd/Counter_t.mli:
```diff
@@ -0,0 +1,10 @@
+(* Auto-generated from "Counter.atd" *)
+              [@@@ocaml.warning "-27-32-35-39"]
+
+type update = { value: int }
+
+type t = { id: string; name: string; value: int }
```
Can you remove this from git?

idkjs replied:
Done.

davesnx commented on package.json:
```diff
@@ -29,6 +29,7 @@
     "taco-ui": "^0.1.2"
   },
   "devDependencies": {
+    "bs-platform": "^9.0.2",
```
Please, remove this.

I want to use melange not bsb

idkjs replied:
Project will not run at this point without it. It will build without error with `esy bsb -make-world` but running `yarn start` will fail because the js build will not find any `bs-platform` related files. It was in the last PR notes. I will re-run it to show you the errors.

idkjs replied:
Tried to reproduce and figured you must think it should work so I went back and found some diffs with the commands in main and now it is working. `bs-platform` is removed.

idkjs commented:
@davesnx I ran `git filter-branch` to remove the generated atdgen files from git history and now I have all these conflicts to resolve. The suggested commands to resolve the conflict have me editing the main branch. Is this something we want to do?

davesnx commented on README.md:
````diff
@@ -1,85 +1,86 @@
+##### Start client
 
+```bash
+yarn start (take into account that Melange crashes on watch mode: https://github.com/melange-re/melange/issues/16)
````
Since we had the watchexec this isn't nececssary
